src/stactools/core/utils/raster_footprint.py

```diff
@@ -13,12 +13,277 @@
+            if not self.asset_names or name in self.asset_names:
+                href = asset.get_absolute_href()
+                if href is None:
+                    logger.error(
+                        f"Could not determine extent for asset '{name}', missing href"
+                    )
+                else:
+                    extent = self.data_footprint(href)
+                    if extent:
+                        yield name, extent
+                    else:
+                        logger.error(f"Could not determine extent for asset '{name}'")
+
+    def data_footprint(self, href: str) -> Optional[Dict[str, Any]]:
+        with rasterio.open(href) as src:
+            self.transform = src.transform
+            self.shape = src.shape
+            if self.crs is None:
+                self.crs = src.crs
+
+            # create datamask
+            if self.no_data is None:
+                self.no_data = src.nodata
+
+            if not src.indexes:
+                raise ValueError(
+                    "Raster footprint cannot be computed for an asset with no bands."
+                )
+            if not self.bands:
+                self.bands = src.indexes
+
+            if self.no_data is not None:
+                self.data: npt.NDArray[np.uint8] = np.full(src.shape, 0, dtype=np.uint8)
+                for index in self.bands:
+                    band_data = src.read(index, out_shape=src.shape)
+                    if np.isnan(self.no_data):
+                        self.data[~np.isnan(band_data)] = 1
+                    else:
+                        self.data[np.where(band_data != self.no_data)] = 1
+            else:
+                self.data = np.full(src.shape, 1, dtype=np.uint8)
+
+        data_polygons = [
+            shape(polygon_dict)
+            for polygon_dict, region_value in rasterio.features.shapes(
+                self.data, transform=self.transform
+            )
+            if region_value == 1
+        ]
+
+        if not data_polygons:
+            return None
+        elif len(data_polygons) == 1:
+            polygon = data_polygons[0]
+        else:
+            polygon = MultiPolygon(data_polygons).convex_hull
+
+        polygon = self.densify(polygon)
+        polygon = self.reproject(polygon)
+        polygon = self.simplify(polygon)
+
+        return mapping(polygon)  # type: ignore
+
+    def densify(self, polygon: Polygon) -> Polygon:
+        raise NotImplementedError
+
+    def reproject(self, polygon: Polygon) -> Polygon:
+        raise NotImplementedError
+
+    def simplify(self, polygon: Polygon) -> Polygon:
+        if self.simplify_tolerance is not None:
+            polygon = polygon.simplify(
+                tolerance=self.simplify_tolerance, preserve_topology=False
+            ).simplify(0)
+
+        # simplify does not remove duplicate sequential points, so do that
+        return Polygon([k for k, _ in groupby(polygon.exterior.coords)])
```

Comment:
So I know this feels like I'm all over the place w/ "free functions vs. class-based", but this method feels likt it could live as its own free function:

```python
def simplify(polygon: Polygon, simplify_tolerance: float) -> Polygon:
    ...
```

This would allow you to remove the `is not None` checks from the function (that could be performed in the class), and would provide the function to folks that just wanted to do simplification w/o all the other rigamarole.

I see below that you've created `_densify_by_factor` as a standalone, "private" function ... a similar `_simplify` function would make sense too, in other words.

Reply:
`simplify()` (now `simplify_polygon()`) is now a single shapely simplify call. I moved the duplicate point removal to where the duplicate points are generated, which is in the reprojection function (that's where the coordinates get rounded, which causes duplicate coords when coarsely rounding). For now I have _not_ extracted a free function from the contents of `reproject_polygon()`.